sdk/test/trace/batch_span_processor_test.cc

```diff
@@ -0,0 +1,344 @@
+        std::shared_ptr<bool> is_shutdown,
+        const std::chrono::milliseconds export_delay = std::chrono::milliseconds(0),
+        const int schedule_delay_millis = 5000,
+        const int max_queue_size = 2048,
+        const int max_export_batch_size = 512
+    )
+    {
+        return std::shared_ptr<sdk::trace::SpanProcessor>(
+            new sdk::trace::BatchSpanProcessor(GetMockExporter(spans_received, is_shutdown, export_delay),
+                                               max_queue_size,
+                                               schedule_delay_millis,
+                                               max_export_batch_size));
+    }
+
+
+    std::unique_ptr<std::vector<std::unique_ptr<sdk::trace::Recordable>>> GetTestSpans(
+        std::shared_ptr<sdk::trace::SpanProcessor> processor,
+        const int num_spans
+    )
+    {
+        std::unique_ptr<std::vector<std::unique_ptr<sdk::trace::Recordable>>> test_spans(
+            new std::vector<std::unique_ptr<sdk::trace::Recordable>>
+        );
+
+        for(int i = 0; i < num_spans; ++i)
+        {
+            test_spans->push_back(processor->MakeRecordable());
+            static_cast<sdk::trace::SpanData*>(test_spans->at(i).get())->SetName("Span " + i);
+        }   
+
+        return test_spans;
+    }
+
+private:
+    std::unique_ptr<sdk::trace::SpanExporter> GetMockExporter(
+        std::shared_ptr<std::vector<std::unique_ptr<sdk::trace::SpanData>>> spans_received,
+        std::shared_ptr<bool> is_shutdown,
+        const std::chrono::milliseconds export_delay = std::chrono::milliseconds(0)
+    )
+    {
+        return std::unique_ptr<sdk::trace::SpanExporter>(new MockSpanExporter(spans_received, is_shutdown, export_delay));
+    }
+};
+
+/* ##################################   TESTS   ############################################ */
+
+TEST_F(BatchSpanProcessorTestPeer, TestShutdown)
+{
+    std::shared_ptr<bool> is_shutdown(new bool(false));
+    std::shared_ptr<std::vector<std::unique_ptr<sdk::trace::SpanData>>> spans_received(
+        new std::vector<std::unique_ptr<sdk::trace::SpanData>>);
+
+    auto batch_processor = GetMockProcessor(spans_received, is_shutdown);
+    const int num_spans = 3;
+
+    auto test_spans = GetTestSpans(batch_processor, num_spans);
+
+    for(int i = 0; i < num_spans; ++i){
+        batch_processor->OnEnd(std::move(test_spans->at(i)));
+    }
+
+    // Wait a bit to add spans to the buffer and for the background worker
+    // thread of the batch processor to wait on a condition variable.
+    // Helps in proper preemption of the worker thread.
+    // NOTE: If this delay is not added, it's possible for Shutdown() to notify the
+    //       worker thread even before it starts waiting, causing it to sleep for
+    //       the entire 'schedule_delay_millis' timeout
+    // TODO: Discuss this
+    std::this_thread::sleep_for(std::chrono::milliseconds(10));
```

Comment:
Please have a look at the explanation for this and let me know if it makes sense.

Reply:
It would be better to rewrite notifications to be more robust, which would allow avoiding this sleep. I think the need for this sleep identifies a problem in the main loop, and masks it.

Reply:
This probably wouldn't be required at all if we used the synchronized Java approach.
Do you recommend making notifications more robust as the first step and later implementing the Java approach, once everyone agrees on its requirements(e.g, changing the exporter interface to include timeout)?